### PR TITLE
Support subquery Criteria API

### PR DESCRIPTION
**Pull request by dstepanov**
First step to support pagination with JOINs

### CODE REVIEW

File: data-model/src/main/java/io/micronaut/data/model/jpa/criteria/PersistentEntitySubquery.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * The persistent entity {@link Subquery}.
+ *
+ * @param <T> The type of the selected item
+ * @author Denis Stepanov
+ * @since 4.10
+ */
+@Experimental
+public interface PersistentEntitySubquery<T> extends Subquery<T>, PersistentEntityCommonAbstractCriteria {
```
Comment (graemerocher) [inline data-model/src/main/java/io/micronaut/data/model/jpa/criteria/PersistentEntitySubquery.java:36]: complete the javadoc for the public methods

File: data-processor/src/main/java/io/micronaut/data/processor/visitors/RepositoryTypeElementVisitor.java [hunk 2]
```diff
@@ -318,6 +318,7 @@ public void visitMethod(MethodElement element, VisitorContext context) {
                 context.fail(matchContext.getUnableToImplementMessage() + e.getMessage(), e.getElement() == null ? element : e.getElement());
                 this.failing = true;
             } catch (Exception e) {
+                e.printStackTrace();
```
Comment (graemerocher) [inline data-processor/src/main/java/io/micronaut/data/processor/visitors/RepositoryTypeElementVisitor.java]: should be removed or handled differently

**Review by graemerocher — CHANGES_REQUESTED** [review]

**Review by radovanradic — APPROVED** [review]

File: data-model/src/main/java/io/micronaut/data/model/jpa/criteria/PersistentEntitySubquery.java (hunk 1 above)
Comment (dstepanov) [reply data-model/src/main/java/io/micronaut/data/model/jpa/criteria/PersistentEntitySubquery.java:36]: Added for new methods, for overriden the javadoc should be taken from the super interface.

File: data-processor/src/main/java/io/micronaut/data/processor/visitors/RepositoryTypeElementVisitor.java (hunk 2 above)
Comment (dstepanov) [reply data-processor/src/main/java/io/micronaut/data/processor/visitors/RepositoryTypeElementVisitor.java]: Uninentialy added

**Review by graemerocher — APPROVED** [review]

**Pull request merged**